File: jas_ocaml/test/algorithms/bristle_stroke_test.ml
(* Equivalence vectors shared with the Rust reference
   (jas_dioxus/src/algorithms/bristle_stroke.rs) and the Swift port: width 4,
   density 25 -> two bristles at ±2 along a straight horizontal path. *)

let brush () : Jas.Bristle_stroke.t =
  Jas.Bristle_stroke.{
    size = 4.0;
    density = 25.0;
    thickness = 30.0;
    opacity = 30.0;
    stroke_weight = 1.0;
  }

let close (ax, ay) (bx, by) =
  Float.abs (ax -. bx) < 1e-6 && Float.abs (ay -. by) < 1e-6

let test_two_bristles () =
  let cmds = [ Jas.Element.MoveTo (0.0, 0.0); Jas.Element.LineTo (100.0, 0.0) ] in
  let out = Jas.Bristle_stroke.stroke cmds (brush ()) in
  Alcotest.(check int) "two bristles" 2 (List.length out);
  let b0 = Array.of_list (List.nth out 0) in
  let b1 = Array.of_list (List.nth out 1) in
  Alcotest.(check bool) "b0 start" true (close b0.(0) (0.0, -2.0));
  Alcotest.(check bool) "b0 end" true (close b0.(1) (100.0, -2.0));
  Alcotest.(check bool) "b1 start" true (close b1.(0) (0.0, 2.0));
  Alcotest.(check bool) "b1 end" true (close b1.(1) (100.0, 2.0))

let test_count_alpha () =
  Alcotest.(check int) "count" 2 (Jas.Bristle_stroke.count (brush ()));
  Alcotest.(check bool) "alpha" true
    (Float.abs (Jas.Bristle_stroke.alpha (brush ()) -. 0.3) < 1e-9)

let () =
  Alcotest.run "BristleStroke"
    [
      ( "stroke",
        [
          Alcotest.test_case "two bristles" `Quick test_two_bristles;
          Alcotest.test_case "count/alpha" `Quick test_count_alpha;
        ] );
    ]

File: jas/algorithms/bristle_stroke_test.py
"""Equivalence vectors shared with the Rust reference
(jas_dioxus/src/algorithms/bristle_stroke.rs) and the Swift / OCaml ports:
width 4, density 25 -> two bristles at +/-2 along a straight horizontal path."""

from algorithms.bristle_stroke import BristleBrush, bristle_stroke
from geometry.element import MoveTo, LineTo


def _brush():
    return BristleBrush(size=4.0, density=25.0, thickness=30.0, opacity=30.0,
                        stroke_weight=1.0)


def _close(a, x, y):
    return abs(a[0] - x) < 1e-6 and abs(a[1] - y) < 1e-6


def test_straight_path_two_offset_bristles():
    out = bristle_stroke([MoveTo(0.0, 0.0), LineTo(100.0, 0.0)], _brush())
    assert len(out) == 2
    assert _close(out[0][0], 0.0, -2.0)
    assert _close(out[0][1], 100.0, -2.0)
    assert _close(out[1][0], 0.0, 2.0)
    assert _close(out[1][1], 100.0, 2.0)


def test_count_and_alpha():
    b = _brush()
    assert b.count() == 2
    assert abs(b.alpha() - 0.3) < 1e-9


def test_empty_for_degenerate():
    assert bristle_stroke([MoveTo(0.0, 0.0)], _brush()) == []
